File: src/modules/sticker_service.py
import os
import json
import subprocess
from pathlib import Path
from typing import List, Dict, Any, Optional, Tuple
from dataclasses import dataclass

from src.config.config import config
from src.utils.helpers import helpers


@dataclass
class ImageInfo:
    width: int
    height: int
    duration: Optional[float]
    is_animated: bool
    format: str


@dataclass
class StickerResult:
    success: bool
    original_video: str
    output_path: str
    filename: str
    sticker: Dict[str, Any]


@dataclass
class MultiStickerResult:
    success: bool
    original_video: str
    output_path: str
    filename: str
    stickers: List[Dict[str, Any]]
    stickers_count: int


@dataclass
class StaticStickerResult:
    success: bool
    original_gif: str
    output_path: str
    filename: str
    frame_index: int


@dataclass
class ValidationResult:
    valid: bool
    type: Optional[str] = None
    width: Optional[int] = None
    height: Optional[int] = None
    is_animated: Optional[bool] = None
    format: Optional[str] = None
    error: Optional[str] = None


class StickerService:
    def __init__(self, options: Optional[Dict[str, Any]] = None):
        options = options or {}
        self.options = {
            'output_dir': options.get('output_dir', config.temp_dir),
            'default_opacity': options.get('default_opacity', config.sticker['default_opacity']),
            'default_scale': options.get('default_scale', config.sticker['default_scale']),
            'default_position': options.get('default_position', config.sticker['default_position']),
            **options,
        }
        self.ffmpeg_path = config.ffmpeg['path']
        self.ffprobe_path = config.ffmpeg['ffprobe_path']

    def get_image_info(self, image_path: str) -> ImageInfo:
        if not os.path.exists(image_path):
            raise Exception(f'图片文件不存在: {image_path}')

        cmd = [
            self.ffprobe_path,
            '-v', 'quiet',
            '-print_format', 'json',
            '-show_format',
            '-show_streams',
            image_path
        ]

        result = subprocess.run(cmd, capture_output=True, text=True)
        if result.returncode != 0:
            raise Exception(f'无法获取图片信息: {result.stderr}')

        metadata = json.loads(result.stdout)
        format_info = metadata.get('format', {})
        streams = metadata.get('streams', [])
        video_stream = next((s for s in streams if s.get('codec_type') == 'video'), None)

        duration = None
        if format_info.get('duration'):
            duration = float(format_info['duration'])

        is_animated = duration is not None and duration > 0

        return ImageInfo(
            width=video_stream.get('width', 0) if video_stream else 0,
            height=video_stream.get('height', 0) if video_stream else 0,
            duration=duration,
            is_animated=is_animated,
            format=format_info.get('format_name', '')
        )

    def _validate_sticker_options(self, sticker: Dict[str, Any], video_info: Dict[str, Any]) -> Dict[str, Any]:
        sticker_path = sticker.get('path')
        sticker_type = sticker.get('type', 'static')
        position = sticker.get('position', self.options['default_position'])
        scale = sticker.get('scale', self.options['default_scale'])
        opacity = sticker.get('opacity', self.options['default_opacity'])
        start_seconds = sticker.get('start_seconds', 0)
        duration = sticker.get('duration')
        x = sticker.get('x')
        y = sticker.get('y')

        if not sticker_path:
            raise Exception('贴纸路径不能为空')

        if not os.path.exists(sticker_path):
            raise Exception(f'贴纸文件不存在: {sticker_path}')

        if scale <= 0:
            raise Exception('缩放比例必须大于0')

        if opacity < 0 or opacity > 1:
            raise Exception('透明度必须在0到1之间')

        if start_seconds < 0:
            raise Exception('开始时间不能为负数')

        if duration is not None and duration <= 0:
            raise Exception('持续时间必须大于0')

        valid_types = ['static', 'gif']
        if sticker_type not in valid_types:
            raise Exception(f'不支持的贴纸类型: {sticker_type}，支持的类型: {", ".join(valid_types)}')

        return {
            'path': sticker_path,
            'type': sticker_type,
            'position': position,
            'scale': scale,
            'opacity': opacity,
            'start_seconds': start_seconds,
            'duration': duration,
            'x': x,
            'y': y,
        }

    def _calculate_position(self, position: str, video_width: int, video_height: int,
                             sticker_width: int, sticker_height: int,
                             x: Optional[float] = None, y: Optional[float] = None) -> Tuple[float, float]:
        if x is not None and y is not None:
            return (x, y)

        x_center = (video_width - sticker_width) / 2
        y_center = (video_height - sticker_height) / 2

        margin = min(video_width, video_height) * 0.05

        positions = {
            'top-left': (margin, margin),
            'top': (x_center, margin),
            'top-right': (video_width - sticker_width - margin, margin),
            'middle-left': (margin, y_center),
            'middle': (x_center, y_center),
            'middle-right': (video_width - sticker_width - margin, y_center),
            'bottom-left': (margin, video_height - sticker_height - margin),
            'bottom': (x_center, video_height - sticker_height - margin),
            'bottom-right': (video_width - sticker_width - margin, video_height - sticker_height - margin),
        }

        return positions.get(position.lower(), positions['top-left'])

    def add_single_sticker(self, video_path: str, sticker: Dict[str, Any],
                             options: Optional[Dict[str, Any]] = None) -> StickerResult:
        options = options or {}
        output_dir = options.get('output_dir', self.options['output_dir'])
        output_filename = options.get('output_filename', f"video_with_sticker_{helpers.generate_unique_id()}.mp4")
        output_path = os.path.join(output_dir, output_filename)

        helpers.ensure_directory(output_dir)

        video_info = self._get_video_info_for_sticker(video_path)

        validated_sticker = self._validate_sticker_options(sticker, video_info)
        sticker_info = self.get_image_info(validated_sticker['path'])

        scaled_width = round(sticker_info.width * validated_sticker['scale'])
        scaled_height = round(sticker_info.height * validated_sticker['scale'])

        x, y = self._calculate_position(
            validated_sticker['position'],
            video_info['width'],
            video_info['height'],
            scaled_width,
            scaled_height,
            validated_sticker['x'],
            validated_sticker['y']
        )

        end_time = validated_sticker['start_seconds'] + validated_sticker['duration'] \
            if validated_sticker['duration'] else video_info['duration']

        gif_loop_count = 1
        gif_actual_duration = sticker_info.duration or 0
        if validated_sticker['type'] == 'gif' and sticker_info.is_animated and gif_actual_duration > 0:
            display_duration = end_time - validated_sticker['start_seconds']
            gif_loop_count = int(display_duration / gif_actual_duration) + 1

        input_args = ['-i', video_path]
        input_options = []

        if validated_sticker['type'] == 'gif' and sticker_info.is_animated:
            input_options.extend(['-ignore_loop', '0'])

        input_args.extend(input_options)
        input_args.extend(['-i', validated_sticker['path']])

        filter_complex = []

        scale_filter = f'[1:v]scale={scaled_width}:{scaled_height}[scaled_sticker]'
        filter_complex.append(scale_filter)

        sticker_input = 'scaled_sticker'

        if validated_sticker['opacity'] < 1:
            alpha = validated_sticker['opacity']
            opacity_filter = f'[scaled_sticker]colorchannelmixer=aa={alpha}[transparent_sticker]'
            filter_complex.append(opacity_filter)
            sticker_input = 'transparent_sticker'

        enable_option = ''
        if validated_sticker['start_seconds'] > 0 or validated_sticker['duration'] is not None:
            start = validated_sticker['start_seconds']
            end = end_time
            enable_option = f":enable='between(t,{start},{end})'"

        overlay_filter = f"[0:v][{sticker_input}]overlay=x={x}:y={y}{enable_option}[out]"
        filter_complex.append(overlay_filter)

        cmd = [self.ffmpeg_path, '-y'] + input_args + [
            '-filter_complex', ';'.join(filter_complex),
            '-map', '[out]',
            '-map', '0:a?',
            '-c:v', 'libx264',
            '-preset', 'medium',
            '-crf', '23',
            '-c:a', 'copy',
            output_path
        ]

        result = subprocess.run(cmd, capture_output=True)
        if result.returncode != 0:
            raise Exception(f'添加贴纸失败: {result.stderr.decode("utf-8", errors="ignore")}')

        sticker_result = {
            **validated_sticker,
            'scaled_width': scaled_width,
            'scaled_height': scaled_height,
            'x': x,
            'y': y,
            'width': sticker_info.width,
            'height': sticker_info.height,
            'gif_duration': gif_actual_duration,
            'gif_loop_count': gif_loop_count,
        }

        return StickerResult(
            success=True,
            original_video=video_path,
            output_path=output_path,
            filename=output_filename,
            sticker=sticker_result
        )

    def add_multiple_stickers(self, video_path: str, stickers: List[Dict[str, Any]],
                                options: Optional[Dict[str, Any]] = None) -> MultiStickerResult:
        if not stickers:
            raise Exception('没有提供贴纸')

        options = options or {}
        output_dir = options.get('output_dir', self.options['output_dir'])
        output_filename = options.get('output_filename', f"video_with_stickers_{helpers.generate_unique_id()}.mp4")
        output_path = os.path.join(output_dir, output_filename)

        helpers.ensure_directory(output_dir)

        video_info = self._get_video_info_for_sticker(video_path)

        validated_stickers = []
        stickers_info = []
        has_gif_stickers = []

        for sticker in stickers:
            validated = self._validate_sticker_options(sticker, video_info)
            info = self.get_image_info(validated['path'])
            validated_stickers.append(validated)
            stickers_info.append(info)
            has_gif_stickers.append(validated['type'] == 'gif' and info.is_animated)

        has_any_gif = any(has_gif_stickers)

        input_args = ['-i', video_path]
        input_options = []

        if has_any_gif:
            input_options.extend(['-ignore_loop', '0'])

        input_args.extend(input_options)

        for sticker in validated_stickers:
            input_args.extend(['-i', sticker['path']])

        filter_complex = []
        current_video = '0:v'

        for i, sticker in enumerate(validated_stickers):
            sticker_info = stickers_info[i]
            input_index = i + 1

            scaled_width = round(sticker_info.width * sticker['scale'])
            scaled_height = round(sticker_info.height * sticker['scale'])

            x, y = self._calculate_position(
                sticker['position'],
                video_info['width'],
                video_info['height'],
                scaled_width,
                scaled_height,
                sticker['x'],
                sticker['y']
            )

            end_time = sticker['start_seconds'] + sticker['duration'] \
                if sticker['duration'] else video_info['duration']

            gif_loop_count = 1
            gif_actual_duration = sticker_info.duration or 0
            if sticker['type'] == 'gif' and sticker_info.is_animated and gif_actual_duration > 0:
                display_duration = end_time - sticker['start_seconds']
                gif_loop_count = int(display_duration / gif_actual_duration) + 1

            scaled_label = f'scaled_{i}'
            scale_filter = f'[{input_index}:v]scale={scaled_width}:{scaled_height}[{scaled_label}]'
            filter_complex.append(scale_filter)

            sticker_label = scaled_label

            if sticker['opacity'] < 1:
                alpha = sticker['opacity']
                transparent_label = f'transparent_{i}'
                opacity_filter = f'[{scaled_label}]colorchannelmixer=aa={alpha}[{transparent_label}]'
                filter_complex.append(opacity_filter)
                sticker_label = transparent_label

            enable_option = ''
            if sticker['start_seconds'] > 0 or sticker['duration'] is not None:
                start = sticker['start_seconds']
                end = end_time
                enable_option = f":enable='between(t,{start},{end})'"

            output_label = f'layer_{i}'
            overlay_filter = f"[{current_video}][{sticker_label}]overlay=x={x}:y={y}{enable_option}[{output_label}]"
            filter_complex.append(overlay_filter)

            current_video = output_label

            validated_stickers[i] = {
                **sticker,
                'scaled_width': scaled_width,
                'scaled_height': scaled_height,
                'x': x,
                'y': y,
                'width': sticker_info.width,
                'height': sticker_info.height,
                'gif_duration': gif_actual_duration,
                'gif_loop_count': gif_loop_count,
            }

        cmd = [self.ffmpeg_path, '-y'] + input_args + [
            '-filter_complex', ';'.join(filter_complex),
            '-map', f'[{current_video}]',
            '-map', '0:a?',
            '-c:v', 'libx264',
            '-preset', 'medium',
            '-crf', '23',
            '-c:a', 'copy',
            output_path
        ]

        result = subprocess.run(cmd, capture_output=True)
        if result.returncode != 0:
            raise Exception(f'添加多个贴纸失败: {result.stderr.decode("utf-8", errors="ignore")}')

        return MultiStickerResult(
            success=True,
            original_video=video_path,
            output_path=output_path,
            filename=output_filename,
            stickers=validated_stickers,
            stickers_count=len(validated_stickers)
        )

    def _get_video_info_for_sticker(self, video_path: str) -> Dict[str, Any]:
        cmd = [
            self.ffprobe_path,
            '-v', 'quiet',
            '-print_format', 'json',
            '-show_format',
            '-show_streams',
            video_path
        ]

        result = subprocess.run(cmd, capture_output=True, text=True)
        if result.returncode != 0:
            raise Exception(f'无法获取视频信息: {result.stderr}')

        metadata = json.loads(result.stdout)
        format_info = metadata.get('format', {})
        streams = metadata.get('streams', [])
        video_stream = next((s for s in streams if s.get('codec_type') == 'video'), None)

        return {
            'width': video_stream.get('width', 1920) if video_stream else 1920,
            'height': video_stream.get('height', 1080) if video_stream else 1080,
            'duration': float(format_info.get('duration', 0)),
        }

    def create_static_sticker_from_gif(self, gif_path: str, frame_index: int = 0,
                                         options: Optional[Dict[str, Any]] = None) -> StaticStickerResult:
        options = options or {}
        output_dir = options.get('output_dir', self.options['output_dir'])
        output_filename = options.get('output_filename', f"static_from_gif_{helpers.generate_unique_id()}.png")
        output_path = os.path.join(output_dir, output_filename)

        helpers.ensure_directory(output_dir)

        if not os.path.exists(gif_path):
            raise Exception(f'GIF文件不存在: {gif_path}')

        gif_info = self.get_image_info(gif_path)
        if not gif_info.is_animated:
            raise Exception('文件不是动画GIF')

        cmd = [
            self.ffmpeg_path, '-y',
            '-i', gif_path,
            '-vf', f'select=eq(n\\,{frame_index})',
            '-vframes', '1',
            output_path
        ]

        result = subprocess.run(cmd, capture_output=True)
        if result.returncode != 0:
            raise Exception(f'从GIF创建静态贴纸失败: {result.stderr.decode("utf-8", errors="ignore")}')

        return StaticStickerResult(
            success=True,
            original_gif=gif_path,
            output_path=output_path,
            filename=output_filename,
            frame_index=frame_index
        )

    def validate_sticker_file(self, sticker_path: str) -> ValidationResult:
        if not os.path.exists(sticker_path):
            return ValidationResult(
                valid=False,
                error=f'文件不存在: {sticker_path}'
            )

        try:
            info = self.get_image_info(sticker_path)

            valid_formats = ['png', 'jpg', 'jpeg', 'gif', 'bmp', 'webp']
            format_valid = info.format and any(f in info.format.lower() for f in valid_formats)

            return ValidationResult(
                valid=format_valid,
                type='gif' if info.is_animated else 'static',
                width=info.width,
                height=info.height,
                is_animated=info.is_animated,
                format=info.format
            )
        except Exception as error:
            return ValidationResult(
                valid=False,
                error=f'无法读取文件: {str(error)}'
            )

File: ts_temp/src/modules/stickerService.js
import ffmpeg from 'fluent-ffmpeg';
import path from 'path';
import fs from 'fs-extra';
import config from '../config/index.js';
import helpers from '../utils/helpers.js';

ffmpeg.setFfmpegPath(config.ffmpeg.path);
ffmpeg.setFfprobePath(config.ffmpeg.ffprobePath);

class StickerService {
  constructor(options = {}) {
    this.options = {
      outputDir: options.outputDir || config.directories.temp,
      defaultOpacity: options.defaultOpacity || config.sticker.defaultOpacity,
      defaultScale: options.defaultScale || config.sticker.defaultScale,
      defaultPosition: options.defaultPosition || config.sticker.defaultPosition,
      ...options,
    };
  }

  async getImageInfo(imagePath) {
    return new Promise((resolve, reject) => {
      ffmpeg.ffprobe(imagePath, (err, metadata) => {
        if (err) {
          reject(new Error(`无法获取图片信息: ${err.message}`));
          return;
        }

        const videoStream = metadata.streams.find(s => s.codec_type === 'video');

        resolve({
          width: videoStream ? videoStream.width : 0,
          height: videoStream ? videoStream.height : 0,
          duration: metadata.format.duration ? parseFloat(metadata.format.duration) : null,
          isAnimated: metadata.format.duration ? parseFloat(metadata.format.duration) > 0 : false,
          format: metadata.format.format_name,
        });
      });
    });
  }

  _validateStickerOptions(sticker, videoInfo) {
    const {
      path: stickerPath,
      type = 'static',
      position = this.options.defaultPosition,
      scale = this.options.defaultScale,
      opacity = this.options.defaultOpacity,
      startSeconds = 0,
      duration = null,
      x = null,
      y = null,
    } = sticker;

    if (!stickerPath) {
      throw new Error('贴纸路径不能为空');
    }

    if (!fs.existsSync(stickerPath)) {
      throw new Error(`贴纸文件不存在: ${stickerPath}`);
    }

    if (scale <= 0) {
      throw new Error('缩放比例必须大于0');
    }

    if (opacity < 0 || opacity > 1) {
      throw new Error('透明度必须在0到1之间');
    }

    if (startSeconds < 0) {
      throw new Error('开始时间不能为负数');
    }

    if (duration !== null && duration <= 0) {
      throw new Error('持续时间必须大于0');
    }

    const validTypes = ['static', 'gif'];
    if (!validTypes.includes(type)) {
      throw new Error(`不支持的贴纸类型: ${type}，支持的类型: ${validTypes.join(', ')}`);
    }

    return {
      stickerPath,
      type,
      position,
      scale,
      opacity,
      startSeconds,
      duration,
      x,
      y,
    };
  }

  _calculatePosition(position, videoWidth, videoHeight, stickerWidth, stickerHeight, x = null, y = null) {
    if (x !== null && y !== null) {
      return { x, y };
    }

    const xCenter = (videoWidth - stickerWidth) / 2;
    const yCenter = (videoHeight - stickerHeight) / 2;

    const margin = Math.min(videoWidth, videoHeight) * 0.05;

    const positions = {
      'top-left': { x: margin, y: margin },
      'top': { x: xCenter, y: margin },
      'top-right': { x: videoWidth - stickerWidth - margin, y: margin },
      'middle-left': { x: margin, y: yCenter },
      'middle': { x: xCenter, y: yCenter },
      'middle-right': { x: videoWidth - stickerWidth - margin, y: yCenter },
      'bottom-left': { x: margin, y: videoHeight - stickerHeight - margin },
      'bottom': { x: xCenter, y: videoHeight - stickerHeight - margin },
      'bottom-right': { x: videoWidth - stickerWidth - margin, y: videoHeight - stickerHeight - margin },
    };

    return positions[position] || positions['top-left'];
  }

  async addSingleSticker(videoPath, sticker, options = {}) {
    const outputDir = options.outputDir || this.options.outputDir;
    const outputFileName = options.outputFileName || `video_with_sticker_${helpers.generateUniqueId()}.mp4`;
    const outputPath = path.join(outputDir, outputFileName);

    helpers.ensureDirectory(outputDir);

    const videoInfo = await new Promise((resolve, reject) => {
      ffmpeg.ffprobe(videoPath, (err, metadata) => {
        if (err) {
          reject(new Error(`无法获取视频信息: ${err.message}`));
          return;
        }
        const videoStream = metadata.streams.find(s => s.codec_type === 'video');
        resolve({
          width: videoStream ? videoStream.width : 1920,
          height: videoStream ? videoStream.height : 1080,
          duration: parseFloat(metadata.format.duration),
        });
      });
    });

    const validatedSticker = this._validateStickerOptions(sticker, videoInfo);
    const stickerInfo = await this.getImageInfo(validatedSticker.stickerPath);

    const scaledWidth = Math.round(stickerInfo.width * validatedSticker.scale);
    const scaledHeight = Math.round(stickerInfo.height * validatedSticker.scale);

    const { x, y } = this._calculatePosition(
      validatedSticker.position,
      videoInfo.width,
      videoInfo.height,
      scaledWidth,
      scaledHeight,
      validatedSticker.x,
      validatedSticker.y
    );

    const endTime = validatedSticker.duration 
      ? validatedSticker.startSeconds + validatedSticker.duration 
      : videoInfo.duration;

    let gifLoopCount = 1;
    let gifActualDuration = stickerInfo.duration || 0;
    if (validatedSticker.type === 'gif' && stickerInfo.isAnimated && gifActualDuration > 0) {
      const displayDuration = endTime - validatedSticker.startSeconds;
      gifLoopCount = Math.ceil(displayDuration / gifActualDuration);
    }

    return new Promise((resolve, reject) => {
      let command = ffmpeg(videoPath);

      if (validatedSticker.type === 'gif' && stickerInfo.isAnimated) {
        command = command.inputOptions('-ignore_loop', '0');
      }

      command = command.input(validatedSticker.stickerPath);

      const filterComplex = [];

      const scaleFilter = `[1:v]scale=${scaledWidth}:${scaledHeight}[scaled_sticker]`;
      filterComplex.push(scaleFilter);

      if (validatedSticker.opacity < 1) {
        const alpha = validatedSticker.opacity;
        const opacityFilter = `[scaled_sticker]colorchannelmixer=aa=${alpha}[transparent_sticker]`;
        filterComplex.push(opacityFilter);
      }

      const stickerInput = validatedSticker.opacity < 1 ? 'transparent_sticker' : 'scaled_sticker';

      let enableOption = '';
      if (validatedSticker.startSeconds > 0 || validatedSticker.duration !== null) {
        const start = validatedSticker.startSeconds;
        const end = endTime;
        enableOption = `:enable='between(t,${start},${end})'`;
      }

      const overlayFilter = `[0:v][${stickerInput}]overlay=x=${x}:y=${y}${enableOption}[out]`;
      filterComplex.push(overlayFilter);

      command.complexFilter(filterComplex, 'out')
        .output(outputPath)
        .outputOptions([
          '-c:v', 'libx264',
          '-preset', 'medium',
          '-crf', '23',
          '-c:a', 'copy',
        ]);

      command
        .on('end', () => {
          resolve({
            success: true,
            originalVideo: videoPath,
            outputPath: outputPath,
            filename: outputFileName,
            sticker: {
              ...validatedSticker,
              scaledWidth,
              scaledHeight,
              x,
              y,
              width: stickerInfo.width,
              height: stickerInfo.height,
              gifDuration: gifActualDuration,
              gifLoopCount: gifLoopCount,
            },
          });
        })
        .on('error', (err) => {
          reject(new Error(`添加贴纸失败: ${err.message}`));
        })
        .run();
    });
  }

  async addMultipleStickers(videoPath, stickers, options = {}) {
    if (!stickers || stickers.length === 0) {
      throw new Error('没有提供贴纸');
    }

    const outputDir = options.outputDir || this.options.outputDir;
    const outputFileName = options.outputFileName || `video_with_stickers_${helpers.generateUniqueId()}.mp4`;
    const outputPath = path.join(outputDir, outputFileName);

    helpers.ensureDirectory(outputDir);

    const videoInfo = await new Promise((resolve, reject) => {
      ffmpeg.ffprobe(videoPath, (err, metadata) => {
        if (err) {
          reject(new Error(`无法获取视频信息: ${err.message}`));
          return;
        }
        const videoStream = metadata.streams.find(s => s.codec_type === 'video');
        resolve({
          width: videoStream ? videoStream.width : 1920,
          height: videoStream ? videoStream.height : 1080,
          duration: parseFloat(metadata.format.duration),
        });
      });
    });

    const validatedStickers = [];
    const stickersInfo = [];
    const hasGifStickers = [];

    for (const sticker of stickers) {
      const validated = this._validateStickerOptions(sticker, videoInfo);
      const info = await this.getImageInfo(validated.stickerPath);
      validatedStickers.push(validated);
      stickersInfo.push(info);
      hasGifStickers.push(validated.type === 'gif' && info.isAnimated);
    }

    const hasAnyGif = hasGifStickers.some(isGif => isGif);

    return new Promise((resolve, reject) => {
      let command = ffmpeg(videoPath);

      if (hasAnyGif) {
        command = command.inputOptions('-ignore_loop', '0');
      }

      for (const sticker of validatedStickers) {
        command = command.input(sticker.stickerPath);
      }

      const filterComplex = [];
      let currentVideo = '0:v';

      for (let i = 0; i < validatedStickers.length; i++) {
        const sticker = validatedStickers[i];
        const stickerInfo = stickersInfo[i];
        const inputIndex = i + 1;

        const scaledWidth = Math.round(stickerInfo.width * sticker.scale);
        const scaledHeight = Math.round(stickerInfo.height * sticker.scale);

        const { x, y } = this._calculatePosition(
          sticker.position,
          videoInfo.width,
          videoInfo.height,
          scaledWidth,
          scaledHeight,
          sticker.x,
          sticker.y
        );

        const endTime = sticker.duration 
          ? sticker.startSeconds + sticker.duration 
          : videoInfo.duration;

        let gifLoopCount = 1;
        let gifActualDuration = stickerInfo.duration || 0;
        if (sticker.type === 'gif' && stickerInfo.isAnimated && gifActualDuration > 0) {
          const displayDuration = endTime - sticker.startSeconds;
          gifLoopCount = Math.ceil(displayDuration / gifActualDuration);
        }

        const scaledLabel = `scaled_${i}`;
        const scaleFilter = `[${inputIndex}:v]scale=${scaledWidth}:${scaledHeight}[${scaledLabel}]`;
        filterComplex.push(scaleFilter);

        let stickerLabel = scaledLabel;

        if (sticker.opacity < 1) {
          const alpha = sticker.opacity;
          const transparentLabel = `transparent_${i}`;
          const opacityFilter = `[${scaledLabel}]colorchannelmixer=aa=${alpha}[${transparentLabel}]`;
          filterComplex.push(opacityFilter);
          stickerLabel = transparentLabel;
        }

        let enableOption = '';
        if (sticker.startSeconds > 0 || sticker.duration !== null) {
          const start = sticker.startSeconds;
          const end = endTime;
          enableOption = `:enable='between(t,${start},${end})'`;
        }

        const outputLabel = `layer_${i}`;
        const overlayFilter = `[${currentVideo}][${stickerLabel}]overlay=x=${x}:y=${y}${enableOption}[${outputLabel}]`;
        filterComplex.push(overlayFilter);

        currentVideo = outputLabel;

        validatedStickers[i] = {
          ...sticker,
          scaledWidth,
          scaledHeight,
          x,
          y,
          width: stickerInfo.width,
          height: stickerInfo.height,
          gifDuration: gifActualDuration,
          gifLoopCount: gifLoopCount,
        };
      }

      command.complexFilter(filterComplex, currentVideo)
        .output(outputPath)
        .outputOptions([
          '-c:v', 'libx264',
          '-preset', 'medium',
          '-crf', '23',
          '-c:a', 'copy',
        ]);

      command
        .on('end', () => {
          resolve({
            success: true,
            originalVideo: videoPath,
            outputPath: outputPath,
            filename: outputFileName,
            stickers: validatedStickers,
            stickersCount: validatedStickers.length,
          });
        })
        .on('error', (err) => {
          reject(new Error(`添加多个贴纸失败: ${err.message}`));
        })
        .run();
    });
  }

  async createStaticStickerFromGif(gifPath, frameIndex = 0, options = {}) {
    const outputDir = options.outputDir || this.options.outputDir;
    const outputFileName = options.outputFileName || `static_from_gif_${helpers.generateUniqueId()}.png`;
    const outputPath = path.join(outputDir, outputFileName);

    helpers.ensureDirectory(outputDir);

    if (!fs.existsSync(gifPath)) {
      throw new Error(`GIF文件不存在: ${gifPath}`);
    }

    const gifInfo = await this.getImageInfo(gifPath);
    if (!gifInfo.isAnimated) {
      throw new Error('文件不是动画GIF');
    }

    return new Promise((resolve, reject) => {
      const command = ffmpeg(gifPath)
        .outputOptions([
          `-vf`, `select=eq(n\\,${frameIndex})`,
          '-vframes', '1',
        ])
        .output(outputPath);

      command
        .on('end', () => {
          resolve({
            success: true,
            originalGif: gifPath,
            outputPath: outputPath,
            filename: outputFileName,
            frameIndex: frameIndex,
          });
        })
        .on('error', (err) => {
          reject(new Error(`从GIF创建静态贴纸失败: ${err.message}`));
        })
        .run();
    });
  }

  async validateStickerFile(stickerPath) {
    if (!fs.existsSync(stickerPath)) {
      return {
        valid: false,
        error: `文件不存在: ${stickerPath}`,
      };
    }

    try {
      const info = await this.getImageInfo(stickerPath);
      
      const validFormats = ['png', 'jpg', 'jpeg', 'gif', 'bmp', 'webp'];
      const formatValid = info.format && validFormats.some(f => info.format.toLowerCase().includes(f));

      return {
        valid: formatValid,
        type: info.isAnimated ? 'gif' : 'static',
        width: info.width,
        height: info.height,
        isAnimated: info.isAnimated,
        format: info.format,
      };
    } catch (error) {
      return {
        valid: false,
        error: `无法读取文件: ${error.message}`,
      };
    }
  }
}

export default StickerService;
